Governance v1

tomhirst opened the pull request:
This PR gets the governance part of the dApp up to a usable v1 level:

- Ensures dApp interacts with the governance smart contracts properly
- Fixes database issues and enriches data with connections
- Borrows some UI from https://compound.finance/governance

Once tested as functional on staging + merged, we can look at adding improvements (with product + design input) in smaller PRs.

**Don't merge prior to testing on Rinkeby**

Some screenshots from my local environment:

![Screenshot 2022-08-12 at 14 50 19](https://user-images.githubusercontent.com/2827412/184368017-48047e9e-1157-4353-915a-feea40270048.png)

![Screenshot 2022-08-12 at 14 50 29](https://user-images.githubusercontent.com/2827412/184368036-a62ce437-dff6-4bb7-8f6e-add7877c6575.png)

![Screenshot 2022-08-12 at 14 50 38](https://user-images.githubusercontent.com/2827412/184368044-8e21e878-2ff9-42e4-9858-3f685dab9310.png)

![Screenshot 2022-08-12 at 14 50 45](https://user-images.githubusercontent.com/2827412/184368073-a9ce085f-8493-4496-8a87-516aa93c51a4.png)

![Screenshot 2022-08-12 at 14 51 02](https://user-images.githubusercontent.com/2827412/184368088-e3c1371f-39a1-448c-9981-0ca2dc99a291.png)

DanielVF commented:
Does this work okay if there is an ampersand or gt or less than in the title? I'm wondering if this is going to be double escaping things.

tomhirst commented:
> Does this work okay if there is an ampersand or > or < than in the title / description? I'm wondering if this is going to be double escaping things.

It does now. Good catch, ser.

![Screenshot 2022-08-12 at 15 02 43](https://user-images.githubusercontent.com/2827412/184370458-6d2d97a3-3217-4d63-aeb9-8051532cc3db.png)

![Screenshot 2022-08-12 at 15 06 02](https://user-images.githubusercontent.com/2827412/184370469-ad62628b-8c71-4f68-a43d-5d4562cb424a.png)

DanielVF commented:
Wearing my paranoid hat here, since an escaping attack on the voting UI would be critical.

I'm concerned that since the `dangerouslySetInnerHTML`'s are in a separate file from the actual escaping, that future changes could inadvertently mess things up.

In `getProposalContent` can we add a comment that the output of this function must be escaped and safe for use as raw HTML in the UI? And could we change the names of in output to append either "safe", "sanitized" or "clean", so that when writing the front end code you know that it's been checked elsewhere.

Ideally we'd have unit tests checking that no scripts could make it through, but we can skip this since we don't have front end browser tests.

tomhirst commented:
> Wearing my paranoid hat here, since an escaping attack on the voting UI would be critical.
> 
> I'm concerned that since the `dangerouslySetInnerHTML`'s are in a separate file from the actual escaping, that future changes could inadvertently mess things up.
> 
> In `getProposalContent` can we add a comment that the output of this function must be escaped and safe for use as raw HTML in the UI? And could we change the names of in output to append either "safe", "sanitized" or "clean", so that when writing the front end code you know that it's been checked elsewhere.
> 
> Ideally we'd have unit tests checking that no scripts could make it through, but we can skip this since we don't have front end browser tests.

Good points. Suggestions added.

tomhirst commented:
Updates from recent feedback post-demo:

- The creation of new proposals from the UI has been disabled
- Truncated addresses now link out to Etherscan

![Screenshot 2022-08-17 at 11 56 32](https://user-images.githubusercontent.com/2827412/185102328-36f384f7-667a-434c-81e0-e0e15b568003.png)

matt-bullock commented:
Will the active tag update to executed when the proposal is completed? It might be good to show a proposal history on the proposal pages.

Could add a back button back to all proposals. I'm used to seeing these on governance UIs.

Could visualise the vote allocation with some kind of bar chart/line diagram like so:
![image](https://user-images.githubusercontent.com/39496225/185159715-cf7269a1-f19f-47a6-87a5-309e7d0eebaf.png)

I'm unable to click on the contract address in the description box.

I'm not really sure what Arguments and Argument types are. Tbh the whole "Description" box doesn't make much sense to me.
![image](https://user-images.githubusercontent.com/39496225/185160047-f17af530-a16a-4cd5-8e8c-7045223bf9cd.png)

Same for the details box. We could look at simplifying this to make it more understandable.

Small UI thing but I struggle to read the text on these disabled buttons. Could up the contrast in the button text so its readable enough whilst still looking disabled. 
![image](https://user-images.githubusercontent.com/39496225/185164152-84fbcc1a-16ef-4d4e-a0fc-6068b8bf5e42.png)

tomhirst commented:
> Will the active tag update to executed when the proposal is completed? It might be good to show a proposal history on the proposal pages.
> 
> Could add a back button back to all proposals. I'm used to seeing these on governance UIs.
> 
> Could visualise the vote allocation with some kind of bar chart/line diagram like so: ![image](https://user-images.githubusercontent.com/39496225/185159715-cf7269a1-f19f-47a6-87a5-309e7d0eebaf.png)
> 
> I'm unable to click on the contract address in the description box.
> 
> I'm not really sure what Arguments and Argument types are. Tbh the whole "Description" box doesn't make much sense to me. ![image](https://user-images.githubusercontent.com/39496225/185160047-f17af530-a16a-4cd5-8e8c-7045223bf9cd.png)
> 
> Same for the details box. We could look at simplifying this to make it more understandable.
> 
> Small UI thing but I struggle to read the text on these disabled buttons. Could up the contrast in the button text so its readable enough whilst still looking disabled. ![image](https://user-images.githubusercontent.com/39496225/185164152-84fbcc1a-16ef-4d4e-a0fc-6068b8bf5e42.png)

Great feedback, thanks @matt-bullock.

1. Yep, the tags update depending on the state of the proposal. Good shout on the history, I've made the relevant changes to the listener to monitor each proposal's state and save timestamps each time it changes.
2. I've added a back button at the top of the single proposals template.
3. Went with the Compound-like progress bars for For/Against voting visualisation.
4. Fixed the contract address not linking out to Etherscan.
5. The description box's purpose is to show two things. 1, the brief description of the proposal submitted on-chain by the user (some of the proposals on RInkeby have no description, which is a little confusing) and 2, to show the actions that will be taken on-chain should the proposal succeed. I'm in agreement that it's not totally clear, though I'm not sure how to improve this visually.
6. Re: details box. We potentially drop this now that we have the history section. Again, I'm not sure how we'd simplify it visually and I'm concerned some users might want to see this information.
7. Good catch on the disabled button text being hard to read. I've increased the contrast there.

From a technical point of view, everything Compound has on screen, we now have. It doesn't look as good as what they have though. Should we invest some time rejigging things in Figma to attain a better layout? Let me know what you think @matt-bullock and @micahalcorn.

tomhirst commented:
![Screenshot 2022-08-19 at 15 09 13](https://user-images.githubusercontent.com/2827412/185637604-43684b84-1bed-4d9e-8eb8-677601e62e01.png)

![Screenshot 2022-08-19 at 15 09 20](https://user-images.githubusercontent.com/2827412/185637615-55a1896a-f972-4249-a411-aaad3fecdc9d.png)

micahalcorn commented:
Video overload 🌊

**P2**
- [ ] [Proposals UX and level of detail](https://www.loom.com/share/1ec4c4e3e6d2426c9e096ae67b13ead7)
- [ ] [Proposal detail confusion](https://www.loom.com/share/37aa4fbcec38426680873d38317c2cc6)
- [ ] [Progress confusion](https://www.loom.com/share/616dde65a26c4f6999daffb10764e5c5)

**P3**
- [ ] [Improve proposal states](https://www.loom.com/share/ef401c790ee34b98bca30ec51aee05ad)
- [ ] [Avoid truncating addresses](https://www.loom.com/share/44790243d2354165b6ac42b29d0493f7)
- [ ] [Support new proposals in test environment](https://www.loom.com/share/36375a8c61734fef8c46a1de9415c103)
- [ ] [Unnecessary Etherscan logos](https://www.loom.com/share/0e82bb2164114380ae9ac2b0fe9c70bd)

tomhirst commented:
> @tomhirst @micahalcorn I'd be happy to work on the design for this. As we have a new brand style/ look and feel which we will soon be introducing, would it would make sense for me to start re-designing it in this style?

Nice! Good point, though. Perhaps my v1 will be good enough from a design perspective until we roll out the wider re-style across Origin products? Though if you can think of any quick layout wins in the current style, let me know and I'll implement them.

matt-bullock commented:
I think this is fine for v1. I look forward to giving it a proper redesign when the time comes. A couple of quick wins that would help in the meantime:

Being able to click the entire area of the proposal to go into it, rather than specific text links. Hopefully this makes sense:

<img width="898" alt="image" src="https://user-images.githubusercontent.com/39496225/186422608-345564e2-e12c-4d18-a7ef-c55a17b0f6ef.png">

---

It would be good to show the for and against percentage bars even if one side has no votes (it would just be the empty grey bar). Like you have here:

<img width="679" alt="image" src="https://user-images.githubusercontent.com/39496225/186423892-29dad36c-7eee-4b5f-880e-44ebd742b1c4.png">


---

Small alignment thing. You can line the text up here so it's all aligned 

<img width="1096" alt="image" src="https://user-images.githubusercontent.com/39496225/186424572-495b3c32-09ed-4785-ba7b-81e151525904.png">

---

For the progress bar in the details section it would help to have the grey bar in the background so show where the current progress sits in relations to being 100% complete. Like you have the grey background in the for and against vote bars.

<img width="587" alt="image" src="https://user-images.githubusercontent.com/39496225/186425101-3776ad07-a7cb-4aec-80df-533f212228e2.png">

tomhirst commented:
> Video overload 🌊
> 
> **P2**
> 
> * [x]  [Proposals UX and level of detail](https://www.loom.com/share/1ec4c4e3e6d2426c9e096ae67b13ead7)
> * [x]  [Proposal detail confusion](https://www.loom.com/share/37aa4fbcec38426680873d38317c2cc6)
> * [x]  [Progress confusion](https://www.loom.com/share/616dde65a26c4f6999daffb10764e5c5)
> 
> **P3**
> 
> * [x]  [Improve proposal states](https://www.loom.com/share/ef401c790ee34b98bca30ec51aee05ad)
> * [ ]  [Avoid truncating addresses](https://www.loom.com/share/44790243d2354165b6ac42b29d0493f7)
> * [x]  [Support new proposals in test environment](https://www.loom.com/share/36375a8c61734fef8c46a1de9415c103)
> * [x]  [Unnecessary Etherscan logos](https://www.loom.com/share/0e82bb2164114380ae9ac2b0fe9c70bd)

Thanks for this @micahalcorn - I've worked through them all and have made some additions to the staging environment.

tomhirst commented:
@micahalcorn Here are a few screenshots from my local testing to add a little colour to the different proposal states:

**Proposal index**

![Screenshot 2022-08-25 at 13 15 32](https://user-images.githubusercontent.com/2827412/186663209-591a220f-b57f-4339-9142-ea4de6d9bb87.png)

**Pending proposal**

![Screenshot 2022-08-25 at 13 16 13](https://user-images.githubusercontent.com/2827412/186663226-ad845232-3cd5-4320-b708-e5cc993c699b.png)

**Active proposal**

![Screenshot 2022-08-25 at 13 16 36](https://user-images.githubusercontent.com/2827412/186663251-6380bb17-6d7a-4097-bfd8-2c911bf94d1c.png)

**Defeated proposal**

![Screenshot 2022-08-25 at 13 17 01](https://user-images.githubusercontent.com/2827412/186663277-9eed2f63-95a1-403f-b61e-63c7cedb0ec6.png)

**Succeeded proposal**
![Screenshot 2022-08-25 at 13 17 22](https://user-images.githubusercontent.com/2827412/186663308-621c4483-d986-4171-8417-d3ab1873d071.png)

Note: In a non-local environment, proposals that succeed will eventually be queued and executed. These events will be picked up by the listener and will further populate the proposal history panel with timestamps and transactions.

micahalcorn commented:
> Note: In a non-local environment, proposals that succeed will eventually be queued and executed. These events will be picked up by the listener and will further populate the proposal history panel with timestamps and transactions.

Amazing, thanks 👏 I'd suggest that we go ahead and proceed. We can continue to test with the new staging contracts and log potential improvements while we move forward as-is with production.

shahthepro commented on client/components/proposal/SupportTable.tsx:
```diff
@@ -0,0 +1,45 @@
+            <td>
+              <TokenAmount
+                amount={ethers.utils.formatUnits(
+                  ethers.BigNumber.from(voter.votes)
```
nit: BigNumber.from can throw if voter.votes isn't BigNumberish

tomhirst replied:
Updated component to catch!

shahthepro commented on client/components/proposal/ProposalVoteStats.tsx:
```diff
@@ -1,78 +1,106 @@
+  votingActive,
 }) => {
+  let tooltipText = "";
+  if (!votePower.gt(0)) {
```
Would be better to check if it's a BigNumber before checking the value

tomhirst replied:
Good catch!

shahthepro commented on client/components/proposal/ProposalTable.tsx:
```diff
@@ -30,58 +37,51 @@ export const ProposalTable = ({ proposalData }) => {
+            const { cleanTitle } = getCleanProposalContent(
+              proposal.description
+            );
+            const id = proposal?.displayId.toString().padStart(3, "0");
```
Any reason why proposal var would be null/undefined? The check is missing in the line above but exists here and on the next one. 

Would it be better to not render the row if it happens to be undefined?

tomhirst replied:
Good spot, there shouldn't be any reason. The optional chaining notations were in the wrong place.

shahthepro commented on client/components/proposal/ProposalTable.tsx:
```diff
@@ -30,58 +37,51 @@ export const ProposalTable = ({ proposalData }) => {
+              <tr
+                key={index}
+                className="cursor-pointer hover:bg-gray-100"
+                onClick={() => router.push(`/proposals/${proposal[0]}`)}
```
proposal is treated as an object above and as an array here, guess it might be a bug

tomhirst replied:
It's actually still an object because it has a `0` key so it's calling in alternative notation. I can see how this is confusing to read though, so I've made it clearer that it's an object using bracket notation.

shahthepro commented on client/components/proposal/ProposalDetail.tsx:
```diff
@@ -95,42 +156,84 @@ export const ProposalDetail = ({
+          </PageTitle>
+          <div className="text-white opacity-70 text-md">
+            {id.toString().padStart(3, "0")} •{" "}
+            {`${lastTx?.event} ${moment(lastTx?.createdAt).format(
```
If lastTx is null, this is gonna show as undefined

tomhirst replied:
Have added a render check to make sure `lastTx` is present.

shahthepro commented on client/components/proposal/ProposalActionsTable.tsx:
```diff
@@ -21,14 +21,14 @@ export const ProposalActionsTable = ({
 
   let explorerPrefix: string | undefined;
-  if (web3Provider?.chainId === 1) {
+  if (rpcProvider?._network?.chainId === 1) {
```
May be, update the hook that returns the provider to also return the chainId? 

This looks a bit hackish because variables and functions prefixed with _ are usually treated as private scoped

tomhirst replied:
Re-written to make it look cleaner.

shahthepro reviewed (commented):
Awesome work, a few comments inline

tomhirst commented on client/utils/useFutureBlockTime.tsx:
```diff
@@ -0,0 +1,30 @@
+      }
+      const res = await fetch(api);
+      const futureBlock = await res.json();
+      if (futureBlock?.status !== "0") {
```
The default state of `futureBlockTime` is 0, so if the call fails, it'll never update. Then in the UI, if this variable is 0, it'll not show the row in the table. Good call on catching the errors - have wrapped in a try/catch.

tomhirst commented on client/utils/index.tsx:
```diff
@@ -150,3 +151,44 @@ export function claimIsOpen() {
+ */
+const getCleanProposalContent = (proposalDescription: string) => {
+  const split = proposalDescription?.split("<br>");
+  const title = split && split[0];
```
Just tested this and `description` returns an empty string if no `<br>` is found in the master string. I think this is OK as we won't render empty vars.

tomhirst commented on client/utils/index.tsx:
```diff
@@ -150,3 +151,44 @@ export function claimIsOpen() {
+  const hoursStr = makeHumanReadable(hours, "hr");
+  const daysStr = makeHumanReadable(days, "day");
+
+  return `${daysStr}${hoursStr}${minutesStr}`.replace(/,\s*$/, "");
```
`trim()` would not go far enough because we need to take away the trailing `,` too.

tomhirst commented on client/pages/proposals/index.tsx:
```diff
@@ -58,12 +61,20 @@ const Proposal: NextPage = ({ proposalCount, proposals }) => {
           displayId: proposals.find(
             (p) => p.proposalId.toString() === d.id.toString()
           )?.id,
+          description: proposals.find(
```
Good catch, this is messy. Have fixed.

tomhirst commented on client/pages/proposals/index.tsx:
```diff
@@ -18,13 +17,17 @@ export async function getServerSideProps({ res }: { res: any }) {
     proposalId: p.proposalId,
     createdAt: p.createdAt.toString(),
+    description: p.description,
+    transactions: JSON.parse(JSON.stringify(p.transactions)),
```
The `createdAt` field on transaction objects can't be serialized as JSON without this, which is a requirement of returning the data in `getServerSideProps`. Have added a comment!

![Screenshot 2022-08-30 at 17 26 54](https://user-images.githubusercontent.com/2827412/187490689-fe962495-e0c1-4af9-9946-b72e0ac51441.png)

tomhirst commented on client/listener.ts:
```diff
@@ -209,21 +293,83 @@ const handleEvents = async (blockNumber, events, done) => {
+                },
+              },
+            });
+            logger.info(`Added ${statusLabels[proposalStatus]} timestamp to proposal ${proposal.proposalId}`);
```
I kind of prefer `info` here to `debug` if following these conventions:

![Screenshot 2022-08-30 at 17 32 58](https://user-images.githubusercontent.com/2827412/187491306-f7c6c224-b050-44c0-a2bd-c19ecd0950f2.png)

Ref: http://www.beefycode.com/post/Log4Net-Tutorial-pt-1-Getting-Started.aspx

tomhirst commented on client/listener.ts:
```diff
@@ -209,21 +293,83 @@ const handleEvents = async (blockNumber, events, done) => {
+      logger.info('No proposals to update.');
+    }
+  } catch(e) {
+    logger.info(e);
```
Agree. Have changed all error messages to `logger.error(e)`.

tomhirst commented:
I gave the "register to vote" CTA some thought overnight and devised this solution. I've re-used an existing CTA pattern to display the section to users with veOGV-positive wallets that haven't self-delegated. This will appear on the home page (last 5 proposals) and the proposals index page. 

LMK if you think anything further is needed @matt-bullock and if you'd like to revise the copy at all @micahalcorn.

<img width="1229" alt="Screenshot 2022-09-02 at 11 00 37" src="https://user-images.githubusercontent.com/2827412/188116023-bd89b96f-0626-4c7e-b44b-b27cc2d758af.png">

<img width="1209" alt="Screenshot 2022-09-02 at 11 00 44" src="https://user-images.githubusercontent.com/2827412/188116055-5960a35c-d3cd-4137-be38-0c065923e665.png">

tomhirst commented:
I've included the CTA on the single proposal view too, thanks for the recommendation @matt-bullock 

<img width="1189" alt="Screenshot 2022-09-02 at 11 46 57" src="https://user-images.githubusercontent.com/2827412/188124735-61a6d20e-a67c-419d-8c0a-569378726771.png">
